### PR TITLE
Firebase clients

### PR DESCRIPTION
**Pull request by norkans7**

### CODE REVIEW

File: core/msgio/android.go [hunk 1]
```diff
@@ -43,22 +45,19 @@ func SyncAndroidChannel(ctx context.Context, fc FCMClient, channel *models.Chann
 
 	start := time.Now()
 
-	if _, err := fc.Send(ctx, sync); err != nil {
+	if _, err := rt.FirebaseCloudMessagingClient.Send(ctx, sync); err != nil {
+		// verify the FCM ID
+		_, err = rt.FirebaseAuthClient.VerifyIDToken(ctx, fcmID)
```
Comment (norkans7) [inline core/msgio/android.go:50]: https://firebase.google.com/docs/auth/admin/verify-id-tokens#go

File: mailroom.go [hunk 2]
```diff
@@ -136,6 +138,17 @@ func (mr *Mailroom) Start() error {
 	// warn if we won't be doing FCM syncing
 	if c.AndroidFCMServiceAccountFile == "" {
 		log.Warn("fcm not configured, no android syncing")
+	} else {
+		app, _ := firebase.NewApp(mr.ctx, nil, option.WithCredentialsFile(c.AndroidFCMServiceAccountFile))
```
Comment (rowanseymour) [inline mailroom.go:142]: is the second returned value here an error and if so why ignore it?

File: testsuite/firebase.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	return &MockFirebaseAuthClient{FirebaseService: m}
+}
+
+func (m *MockFirebaseService) GetAuthClient(ctx context.Context) *MockFirebaseAuthClient {
+	return m.AuthClient(ctx, "testfiles/android.json")
+}
+
+func NewMockFirebaseService(tokens ...string) *MockFirebaseService {
+	mock := &MockFirebaseService{tokens: tokens}
+	return mock
```
Comment (rowanseymour) [inline testsuite/firebase.go:38]:
```
return &MockFirebaseService{tokens: tokens}
```

File: testsuite/testsuite.go [hunk 4]
```diff
@@ -135,6 +138,19 @@ func getES() *elasticsearch.TypedClient {
...
+func getFirebaseAuthClient(ctx context.Context) *MockFirebaseAuthClient {
+	mockFirebase := NewMockFirebaseService("FCMID3")
+	fc := mockFirebase.GetAuthClient(ctx)
+	return fc
+}
+
+func getFirebaseCloudMessagingClient(ctx context.Context) *MockFirebaseCloudMessagingClient {
+	mockFCM := NewMockFirebaseService("FCMID3")
+	fc := mockFCM.GetFirebaseCloudMessagingClient(ctx)
+	return fc
```
Comment (rowanseymour) [inline testsuite/testsuite.go:150]:
```
return mockFCM.GetFirebaseCloudMessagingClient(ctx)
```

**Review by rowanseymour — COMMENTED** [review]
I think your abstraction of FCM functionality to allow mocking requires a bit more thought.. can we have a PR with just the endpoint addition?

**Comment by norkans7** [comment]
@rowanseymour So I add that endpoint without tests for now?

**Comment by rowanseymour** [comment]
Sure

**Comment by norkans7** [comment]
Replaced by https://github.com/nyaruka/mailroom/pull/294

**Pull request closed without merging**